### DISCUSSION TITLE
Device Tracking keeps falling off

### DISCUSSION THREAD

**Discussion by phoniclynx** [mKeRix/room-assistant#1103] [Q&A]
I have been playing around with room-assistant and I've got a few devices with a few quirks and it almost destroys the advantage of using the software for me.

I have 4 devices I am trying to track: Samsung Active 2 Watch. Samsung Active 4 Watch, Tile square and Tile Credit Card. Each seems to have its oddities, and I can't work my way around it.

First my setup. I have 3 devices. Raspberry Pi 4 running Home Assistant Core, Raspberry Pi 3b+, Raspberry Pi 4 all in different rooms and set up as a cluster.

Each device has its sittings appropriate to them:
```
global:
  instanceName: LivingRoom [MyRoom, Kitchen]
  integrations:
    - bluetoothLowEnergy
    - homeAssistant
homeAssistant:
  mqttUrl: mqtt://192.x.x.x:1883
  mqttOptions:
    username: ro....t
    password: 5.....E
  discoveryPrefix: homeassistant

cluster:
  port: 6425
  weight: 100 [75/50]
  peerAddresses:
#    - 192.168.x.x:6425
    - 192.168.x.x:6425
    - 192.168.x.x:6425
    - 192.168.x.x:6425
```

All devices have a list of devices as such:
```
bluetoothLowEnergy:
  timeout: 60
  maxDistance: 3
  hciDeviceId: 0
  allowlist:
    - c0f542750688
    - 'Galaxy Watch4 (CBAA)'
    - 6a1d37636939
    - 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783
    - 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158
  tagOverrides:
    6a1d37636939:
      name: Denises Watch
    Galaxy Watch4 (CBAA):
      name: Denises Watch Name
    c0f542750688:
      name: Nats Watch
    9d41000035d6f4ddba60e7bd8dc491c0-20492-23783:
      name: Nats Tile Keys
      measuredPower: -42
    9d41000035d6f4ddba60e7bd8dc491c0-48300-7158:
      name: Nats Tile Credit Card
````

[Galaxy Watch 4]
I have found it really hard to find the MAC address for the Watch as the MAC would continuously be different, but it would occasionally display its name at the beginning (captured from logs when could find it):
```
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch4 (CBAA) with ID 4c40b0b5fbd7 and RSSI -86
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch4 (CBAA) with ID 5001dc6757ce and RSSI -93
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch4 (CBAA) with ID 693bca9cba46 and RSSI -89
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch4 (CBAA) with ID 4184ef613f87 and RSSI -77
````
It didn't seem to matter how I put it into the config file it wouldn't be picked up. I tried with single quotes, double quotes, no quotes and the software couldn't pick it up as a device. I couldn't use the MAC address as there wasn't a consistent address to use.

[Galaxy Watch 2]
This is an older version of the above watch. It doesn't seem to have the cycling of the MAC address but the address where the name shows up is sometimes different from the one that is listed above in the config files:
```
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch Active2(0344) LE with ID c0f542750688 and RSSI -56
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral Galaxy Watch Active2(0344) LE with ID c0f542750688 and RSSI -61
````
Using this MAC address, however, I can seem to track it. But it has a different quirk. If I turn the watch to Airport Mode and back on it will show up in the logs as above. It then connects to the mobile phone and then it disappears from the logs and presence system. It doesn't update the distance and then after a very short while it displays as "not_home". If I turn off the Bluetooth of the mobile phone the devices mostly seem to work as I expect and they can be picked up by the various Clusters. As soon as I turn the Bluetooth back on room-assistant can't seem to see the watch any more.

[Tile Tags] - both
These seem to have a very similar issue. They seem to react in the same way as the Galaxy Watch 2 where if you have it connected to the Tile App on the mobile it just disappears from room-assistant. The caveat of these devices is that it only seems to show up when you press the button on the Tile itself, which is normally used to locate the mobile phone through the Tile app. If you remove the app or turn off Bluetootn then connect it in one location move it to another far away location in the same cluster group / different raspberry pi, it doesn't update. No distance update and no room presence updates.
```
[Tile Keys]
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 with ID 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 and RSSI -82
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 with ID 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 and RSSI -64
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 with ID 9d41000035d6f4ddba60e7bd8dc491c0-20492-23783 and RSSI -51
```
```
[Tile Credit Card]
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 with ID 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 and RSSI -60
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 with ID 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 and RSSI -60
info - BluetoothLowEnergyService: Discovered nearby BLE peripheral 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 with ID 9d41000035d6f4ddba60e7bd8dc491c0-48300-7158 and RSSI -55

```

Unfortunately, this means they are all not very useful to me. The Tile app does update the Tile and communicates with the Tile inside the app and can tell me where it is and how far away it is, but room-assistant doesn't keep in contact with the Tile after it's initial discovery.

I did notice that I got this error:
```
error - BluetoothService: Failed to start advertising instance via BLE: Command Disallowed
```

How can I fix these issues?

**Comment by phoniclynx** [comment]
I have found an answer to the Galaxy 2 creating a contestant iBeam. In the Samsung Store, there is a program "Beacon Maker" which creates an iBeacon that runs in the background on the watch and connects to room-assistant. 
Having said that I am having trouble with one of the Raspberry Pi's where if I stop and start the service, it says it is the closest and the device is in this room I already am in. 
However without moving the status changes to "not_home" after a few mins. And then occasionally it will come back and then disappear again.

```
    f2d3fa1534f5000fc2257b7c1140ffc0-0-0:
      name: Nats iBeacon Watch
      measuredPower: -35
```
```
bluetoothLowEnergy:
  timeout: 60
  maxDistance: 500
  hciDeviceId: 0
  onlyIBeacon: true
```

I have noticed that the distance measurements are also widely off.. sometimes its 5 then other times its over 600 when the device hasn't moved.

**Comment by phoniclynx** [comment]
![image](https://user-images.githubusercontent.com/3883152/158277934-e493486c-c925-4aeb-84fa-e676e281f138.png)
This is getting frustrating.. After a bit of work, I can get the 3 in the cluster to kind of respond. I have put a new SDCard into "NatsRoom" and reinstalled system from scratch. The distance of all of the systems went down from over 400 to just under 1. This seemed to make a change on ALL of the devices. Not sure why

Anyhow the next morning has come and I've let them run all night. The beacon software on the watch has been going constantly all night long. I walked into the room where "NatsRoom" was and Area Number 1 it showed up for a short amount of time. I left for a bit and came back and stayed for the duration of Area Number 2. The whole of that time I hadn't moved as I was sitting in a chair.. At point of Area Number 3 I actually decided to stop and start the service for "NatsRoom". The iBeacon showed up for a short while then said "not_home" again and all that time I hadn't moved from that spot. When I did look at the distance for it when I had reset the service it was sitting around 0.2. 

I don't or can't see any other service or diagnostics to see what the servers are thinking because it displays as everything is not here. I don't know where to go from here other than to buy some more SD cards because it's possible they are just old and faulty and start again.

Log of NatsRoom:
```
Mar 14 23:15:27 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:27 - info - ConfigService: Loading configuration from /usr/lib/node_modules/room-assistant/dist/config/definitions/default.js, config/local.yml (Current: /home/pi)
Mar 14 23:15:28 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:28 - info - HomeAssistantService: Successfully connected to MQTT broker at mqtt://192.168.x.x:1883
Mar 14 23:15:28 room-assistant-natsroom systemd[1]: Started room-assistant service.
Mar 14 23:15:28 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:28 - info - NestApplication: Nest application successfully started
Mar 14 23:15:28 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:28 - info - ClusterService: Added 192.168.x.x:6425 to the cluster with id livingroom
Mar 14 23:15:28 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:28 - info - ClusterService: Added 192.168.x.x:6425 to the cluster with id kitchen
Mar 14 23:15:29 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:29 - info - BluetoothLowEnergyService: Discovered nearby BLE peripheral f2d3fa1534f5000fc2257b7c1140ffc0-0-0 with ID f2d3fa1534f5000fc2257b7c1140ffc0-0-0 and RSSI -51
Mar 14 23:15:29 room-assistant-natsroom room-assistant[2238]: 14/03/2022, 23:15:29 - info - BluetoothLowEnergyService: Discovered nearby BLE peripheral d1338ace002d44af88d1e57c12484966-1-19616 with ID d1338ace002d44af88d1e57c12484966-1-19616 and RSSI -82

```

Changes to the configs (taken out irrelevant iBeacons):
```
bluetoothLowEnergy:
  maxDistance: 2.5
  hciDeviceId: 0
  onlyIBeacon: true
  timeout: 3
  allowlist:
    - c9f0363d55ee42b481c0219ec6c9d0a7-0-0
  tagOverrides:
    f2d3fa1534f5000fc2257b7c1140ffc0-0-0:
      name: Nats iBeacon Watch

```

HELP! I'm really not sure where to head now?

**Reply by mKeRix** [reply]
That is certainly strange behavior! Unfortunately, I can't really see a pattern in what you described that would pinpoint to any specific issue. If you have a separate phone that you can install a BLE scanner app on, I would maybe take a look at that and make sure that the Watch is constantly broadcasting under the same iBeacon id.

Within room-assistant you can get a bit more information in two different spots. The [API](https://www.room-assistant.io/guide/api.html) on each instance will export more detailed state data for the entities, including the last recorded measurements per instance. On top of that, each instance will expose [metrics](https://www.room-assistant.io/guide/monitoring.html#metrics) in Prometheus format. You could take a look at `http://<url-to-your-ra-instance>:6415/metrics` by hand as well, even if it isn't particularly friendly to the eye. There is a counter that is continually increasing for each BLE advertisement it receives for example, which could be useful to see if the Bluetooth scanning on the Pi is stalling out. (Usually you would receive warning messages in the logs for these cases though.)

**Reply by worm5406** [reply]
How would one tell if stalling out?

Any other items after the port to help troubleshoot? \ble \network \stats ???
